### ISSUE TITLE
How to implement custom CandidateSpace?

### ISSUE THREAD

**Issue by chaturv3di** [HazyResearch/snorkel#1055]
**TL;DR**

What's the best way to implement a new `CandidateSpace` class? Could someone please provide a list of steps, even if at a high level, that I could follow?

----
**Detail**

It seems to me that, at present, Ngrams are the only candidates that are supported. The `apply` method therein yields `TemporarySpan` objects that are tightly coupled with `Sentence` objects. That seems to be a problem for me. In particular, I'm interested in using entire documents as candidates [see #1051]. At present, I'm manually stitching all the sentences into a single one and faking a `TemporarySpan` object by passing the stitched sentence as the `sentence` parameter. I've included the class definition in the issue linked above.

Although I can `yield` full texts of documents as `TemporarySpan` objects in this manner, for some reason, `session.add()` on this full text object  isn't working properly. I say this because `session.query(Candidate).all()` yields single-sentence spans. These single sentences are chosen seemingly at random from the original full-texts. I am unable to decide how to proceed.

In particular:

1. Must the `apply` method of any new `CandidateSpace` yield objects of a class that is derived from `TemporarySpan`?
2. Alternatively, does it suffice to sub-class `TemporaryContext`? This is not so clear since `TemporarySpan` has many more methods than `TemporaryContext`, and I'm not sure if not implementing these methods would break things downstream.
3. Do I have to create some configs for SQLAlchemy table for these candidates? Where to I place this configuration?
4. Anything else? 

I am not sure if this makes complete sense, but I'll be happy to answer clarifying questions. Any help would be appreciated.

**Comment by chaturv3di** [comment]
Okay, I've narrowed down the scope of this query. I now have implemented a custom `TemporaryContext` sub-class and a corresponding `CandidateSpace` sub-class.

```python
class TemporaryDocument(TemporaryContext):
    def __init__(self, document):
        super(TemporaryDocument, self).__init__()
        self.document = document
        self.name = document.name
        self.meta = document.meta
        return
    
    def __eq__(self, other):
        return self.document.id == other.id
    
    def __ne__(self, other):
        return not(self.__eq__(other))
    
    def __hash__(self):
        return hash(self.document.id)
    
    def get_stable_id(self):
        return self.document.id
    
    def _get_table_name(self):
        return 'document'
    
    def _get_polymorphic_identity(self):
        return 'document'

    # Still a bit problematic
    def _get_insert_query(self):
        return """INSERT INTO document VALUES(:id, :name, :meta)"""

    # Still a bit problematic, same as the above
    def _get_insert_args(self):
        return {'id': self.id, 'name': self.name, 'meta': self.meta}
    
    def __repr__(self):
        return self.document.__repr__()

class DocCandidate(CandidateSpace):
    def apply(self, docContext):
        yield TemporaryDocument(docContext)
```

So now, the candidate extraction process looks like this:

```python
fullDoc = DocCandidate()
defaultMatcher = Matcher()
cand_extractor = CandidateExtractor(Scam, [fullDoc], [defaultMatcher])
```

But now, I'm receiving an SQLAlchemy error regarding malformed `INSERT` query. 
```python
InterfaceError: (sqlite3.InterfaceError) Error binding parameter 2 - probably unsupported type. [SQL: 'INSERT INTO document VALUES(?, ?, ?)'] [parameters: (1069, 'Document-00038', {'file_name': 'allData500.tsv', 'record_num': 38})] (Background on this error at: http://sqlalche.me/e/rvf5)
```

It seems that this only has to do with `_get_insert_query` and `_get_insert_args` methods of the `TemporaryDocument` class. I hope that I am on the right track and that resolving this will let me proceed with the rest of the pipeline as demonstrated in the tutorials.

**Comment by chaturv3di** [comment]
Update: I seem to have resolved this. Just added the statement `self.id = document.id` in the constructor of `TemporaryDocument`. This also solves #1039.

**Comment by highflyingbird-ust** [comment]
Great solution! I needed to use sentences as candidate spans. I was trying to build matcher and span sub classes and decided to go through the issues. Thanks!
